File: history/day_24.ts
import { Logger, Part, run, Type } from "../day_utils"
import { ExtendedMap } from "../mapUtils";
import { PriorityQueue, QueuedItem } from "../priority_queue";

enum ObjectType {
    BLIZZARD_DOWN = "v",
    BLIZZARD_UP = "^",
    BLIZZARD_LEFT = "<",
    BLIZZARD_RIGHT = ">",
    WALL = "#",
}

type CellContent = ObjectType | undefined;


interface World {
    width: number;
    height: number;
    cells: CellContent[][]
}



function parse(lines: string[]): World {
    const cells = lines.map((line, y) => {
        const cells = line.split("").map((char, x) => {
            if (char === ".") {
                return undefined
            }
            return char as ObjectType
        });
        cells.shift();
        cells.pop();
        return cells;
    });
    cells.shift();
    cells.pop();
    return {
        width: cells[0].length,
        height: cells.length,
        cells
    };
}

interface Coord {
    x: number,
    y: number
}

function isOccupied(coord: Coord, min: number, world: World): boolean {
    if (coord.y < 0 || coord.y >= world.height) {
        return false;
    }
    let effectiveXRight = (coord.x - min) % world.width;
    if (effectiveXRight < 0) { effectiveXRight += world.width; }
    const effectiveXLeft = (coord.x + min) % world.width;
    let effectiveYDown = (coord.y - min) % world.height;
    if (effectiveYDown < 0) { effectiveYDown += world.height; }
    const effectiveYUp = (coord.y + min) % world.height;

    return world.cells[coord.y][effectiveXLeft] === ObjectType.BLIZZARD_LEFT ||
        world.cells[coord.y][effectiveXRight] === ObjectType.BLIZZARD_RIGHT ||
        world.cells[effectiveYDown][coord.x] === ObjectType.BLIZZARD_DOWN ||
        world.cells[effectiveYUp][coord.x] === ObjectType.BLIZZARD_UP;
}

interface State {
    coord: Coord,
    mins: number
}

const offsets = [
    {
        offsetX: -1,
        offsetY: 0
    },
    {
        offsetX: 1,
        offsetY: 0
    },
    {
        offsetX: 0,
        offsetY: -1
    },
    {
        offsetX: 0,
        offsetY: 1
    }, {
        offsetX: 0,
        offsetY: 0
    }
]

function nextPossibleStates(state: State, world: World): State[] {
    const newMinutes = state.mins + 1;
    const newStates: State[] = [];
    for (const { offsetX, offsetY } of offsets) {
        const newCoord: Coord = { x: state.coord.x + offsetX, y: state.coord.y + offsetY };
        const isInternalCoord = newCoord.x >= 0 && newCoord.x < world.width &&
            newCoord.y >= 0 && newCoord.y < world.height;
        const isStartingCoord = (newCoord.x === 0 && newCoord.y === -1) ||
            (newCoord.y === world.height && newCoord.x === world.width - 1);
        if ((isInternalCoord || isStartingCoord) &&
            !isOccupied(newCoord, newMinutes, world)) {
            newStates.push({
                mins: newMinutes,
                coord: newCoord
            })
        }
    }
    return newStates;
}

function findMinPath(world: World, start: Coord, target: Coord, startMin: number): number {
    const priorityQueue = new PriorityQueue<State>((s) => s.mins + Math.abs(target.x - s.coord.x) + Math.abs(target.y - s.coord.y));
    const calcKey = (s: State) => `${s.coord.x}:${s.coord.y}|${s.mins}`;
    const initState = {
        coord: start,
        mins: startMin
    }
    priorityQueue.put(initState, calcKey(initState));
    let queuedItemToExplore: QueuedItem<State> | undefined;
    while ((queuedItemToExplore = priorityQueue.pop()) !== undefined) {
        const state = queuedItemToExplore.item;
        if (state.coord.x === target.x && state.coord.y === target.y) {
            return state.mins;
        }
        const newStates = nextPossibleStates(state, world);
        for (const newState of newStates) {
            priorityQueue.put(newState, calcKey(newState));
        }
    }

    throw new Error("Shouldn't occurs");
}



function puzzle(lines: string[], part: Part, type: Type, logger: Logger): void {
    const world = parse(lines);
    const start = { x: 0, y: -1 };
    const end = { x: world.width - 1, y: world.height };
    const firstTrip = findMinPath(world, start, end, 0);
    const secondTrip = findMinPath(world, end, start, firstTrip);
    const thirdTrip = findMinPath(world, start, end, secondTrip);
    logger.result([firstTrip, thirdTrip], [18, 292, 54, 816])
}

run(24, [Type.TEST, Type.RUN], puzzle, [Part.ALL])

File: rust/src/day24.rs
use crate::{
    check_result,
    priority_queue::{Cost, Key, PriorityQueue},
    utils::Context,
};

#[derive(Debug, PartialEq, Eq)]
enum ObjectType {
    BlizzardDown,
    BlizzardUp,
    BlizzardLeft,
    BlizzardRight,
    Empty,
}

//type CellContent = ObjectType | undefined;

struct World {
    width: i16,
    height: i16,
    cells: Vec<ObjectType>,
}

fn parse(lines: &Vec<String>) -> World {
    let height: i16 = lines.len() as i16 - 2;
    let width: i16 = lines.get(0).map(|line| line.len() as i16 - 2).unwrap_or(0);
    World {
        width,
        height,
        cells: lines
            .into_iter()
            .take(lines.len() - 1)
            .skip(1)
            .flat_map(|line| {
                line.chars()
                    .into_iter()
                    .take(line.len() - 1)
                    .skip(1)
                    .map(|c| match c {
                        '^' => ObjectType::BlizzardUp,
                        'v' => ObjectType::BlizzardDown,
                        '>' => ObjectType::BlizzardRight,
                        '<' => ObjectType::BlizzardLeft,
                        _ => ObjectType::Empty,
                    })
            })
            .collect(),
    }
}

#[derive(Debug)]
struct Coord {
    x: i16,
    y: i16,
}

#[derive(Debug)]
struct State<'a> {
    coord: Coord,
    mins: u16,
    target: &'a Coord,
}

static OFFSET: &'static [(i16, i16)] = &[(-1, 0), (1, 0), (0, -1), (0, 1), (0, 0)];

fn calc_index(coord: &Coord, world: &World, minutes_offset: (i16, i16)) -> usize {
    if minutes_offset.1 == 0 {
        let mut offset = (coord.x + minutes_offset.0) % world.width;
        if offset < 0 {
            offset += world.width
        }
        (coord.y * world.width + offset) as usize
    } else {
        let mut offset = (coord.y  + minutes_offset.1) % world.height;
        if offset < 0 {
            offset += world.height
        }
        (offset * world.width + coord.x) as usize
    }
}

fn is_occupied(coord: &Coord, min: &u16, world: &World) -> bool {
    let d: i16 = *min as i16;
    return world.cells[calc_index(coord, world, (d, 0))] == ObjectType::BlizzardLeft
        || world.cells[calc_index(coord, world, (-d, 0))] == ObjectType::BlizzardRight
        || world.cells[calc_index(coord, world, (0, -d))] == ObjectType::BlizzardDown
        || world.cells[calc_index(coord, world, (0, d))] == ObjectType::BlizzardUp;
}

fn get_next_state<'a>(curr: &State<'a>, world: &World) -> Vec<State<'a>> {
    OFFSET
        .into_iter()
        .map(|offset| State {
            coord: Coord {
                x: curr.coord.x + offset.0,
                y: curr.coord.y + offset.1,
            },
            mins: curr.mins + 1,
            target: curr.target,
        })
        .filter(|new_state| {
            let is_start = new_state.coord.y == -1 && new_state.coord.x == 0;
            let is_end = new_state.coord.y == world.height
                && new_state.coord.x == (world.width - 1);
            let is_valid_coord = (new_state.coord.x >= 0 && new_state.coord.x < world.width)
                && (new_state.coord.y >= 0 && new_state.coord.y < world.height);
            is_start
                || is_end
                || (is_valid_coord && !is_occupied(&new_state.coord, &new_state.mins, world))
        })
        .collect()
}

impl<'a> Cost<u16> for State<'a> {
    fn cost(&self) -> u16 {
        self.mins
            + (self.target.x - self.coord.x).abs() as u16
            + (self.target.y - self.coord.y).abs() as u16
    }
}

impl<'a> Key<u32> for State<'a> {
    fn key(&self) -> u32 {
        (self.mins as u32) << 16 | (((self.coord.y + 1) as u32) << 8) | (self.coord.x + 1) as u32
    }
}

fn find_path<'a>(world: &World, start: &Coord, target: &'a Coord, start_min: &u16) -> Option<u16> {
    let mut priority_queue: PriorityQueue<u16, u32, State<'a>> = PriorityQueue::new();
    priority_queue.push(State {
        coord: Coord {
            x: start.x,
            y: start.y,
        },
        mins: *start_min,
        target,
    });
    while let Some(next) = priority_queue.pop() {
        if next.coord.x == target.x && next.coord.y == target.y {
            return Option::Some(next.mins);
        }
        for new_state in get_next_state(&next, &world) {
            priority_queue.push(new_state)
        }
    }
    return Option::None;
}

pub fn puzzle(context: &Context, lines: &Vec<String>) {
    let world = parse(lines);
    let start = Coord { x: 0, y: -1 };
    let end = Coord {
        y: world.height as i16,
        x: world.width as i16 - 1,
    };

    let first_trip = find_path(&world, &start, &end, &0).unwrap();
    let second_trip = find_path(&world, &end, &start, &first_trip).unwrap();
    let third_trip = find_path(&world, &start, &end, &second_trip).unwrap();
    check_result!(context, [first_trip, third_trip], [18, 292, 54, 816]);
}
